pyro/distributions/multivariate_normal.py

```diff
@@ -29,15 +29,30 @@ def linear_solve_compat(matrix, matrix_chol, y):
 
 def matrix_inverse_compat(matrix, matrix_chol):
     """Computes the inverse of a positive semidefinite square matrix."""
-    assert matrix.requires_grad == matrix_chol.requires_grad
-    if matrix.requires_grad:
+    if matrix.requires_grad or matrix_chol.requires_grad:
         # If derivatives are required, use the more expensive inverse.
         return torch.inverse(matrix)
     else:
         # Use the cheaper Cholesky based potri.
         return potri_compat(matrix_chol)
 
 
+class _NormalizationConstant(Function):
+    @staticmethod
+    def forward(ctx, sigma, sigma_cholesky, inverse, dimension, return_zero):
+        ctx.save_for_backward(inverse, sigma_cholesky)
+        results = torch.Tensor(
+            [torch.log(sigma_cholesky.diag()).sum() + (dimension / 2) * np.log(2 * np.pi)]).type_as(sigma_cholesky)
+        return results if not return_zero else torch.zeros(1).type_as(sigma_cholesky)
```

Comment:
I would recommend (1) preserving the parameter name `normalized` throughout, and (2) avoiding unnecessary computation
```py
if not normalized:
    return torch.zeros(1).type_as(sigma_cholesky)
return sigma_cholesky.diag().log().sum() + ...  # no need for .type_as() here
```

Reply:
I will split the if statement, but `return sigma_cholesky.diag().log().sum()` sadly does not work. The `.sum()` does not return a Tensor but a float (or whatever the type is) and pytorch won't accept that return type. That is the reason for the `torch.Tensor(...).type_as`.

Reply:
Gah, I forgot about that `.sum()` behavior, but I'm not sure your version supports batching. What about this?
```py
sigma_cholesky.diag().log().sum(0) + ...
```

Reply:
Hmm, even `.diag()` doesn't support batching...

Reply:
That seems to work.

Reply:
Yeah, but this whole thing does not support batched parameters, so that does not really matter for now.